src/asm.rs

```diff
@@ -6,14 +6,15 @@ use crate::asm;
 #[inline(always)]
 pub fn nop() {
     unsafe {
-        asm!("nop");
+        asm!("nop", options(nostack, preserves_flags));
```

Comment:
The optimization we want to prevent here is the removal of asm (nop), right?
If so, perhaps adding `nomem` and a comment like "Do not use `pure` because prevent nop from being removed" would be preferable.

Reply:
If I'm using `nop`, I'm probably using it for timing purposes. For timing between I/O reads/writes, which are inherently volatile (and thus won't be moved around the `nop`) there is no problem. But if I'm trying to time between doing an algorithm and an I/O read/write, I don't want the end of the algorithm to be moved after the `nop`.

So I think I want to hold off on adding `nomem`. The remaining flags look fine to me, I just want to test locally and see what happens w/ a test firmware :).

Reply:
> But if I'm trying to time between doing an algorithm and an I/O read/write, I don't want the end of the algorithm to be moved after the `nop`.

Hmm, I wonder if it would be preferable to explicitly use barriers in this use case.


Also, [cortex-m crate seems to use `nomem` in nop](https://github.com/rust-embedded/cortex-m/blob/4e908625204a1e95dd3fd5bdcd8d66d6bc11c3bc/src/asm.rs#L54).

Reply:
>I wonder if it would be preferable to explicitly use barriers in this use case.

Inside the `nop()` `fn` implementation, or tell the user to add them manually when calling `nop()`?

Reply:
I meant the latter.

Reply:
Barrier looks like it does what `nop` does just without `nomem`. I think for something like `nop` we should follow what `cortex-m` does and use `barrier` for timing usecases. We should also document this distinction.

Reply:
>I think for something like nop we should follow what cortex-m does and use barrier for timing usecases.

Fair enough; if the `cortex-m` ppl change `nop` behavior later, we can follow suit.

>We should also document this distinction.

@taiki-e Would you be willing to add this documentation w/ an example, and then I'll accept this? Thanks!

Reply:
@taiki-e Did you get the chance to update this PR w/ documentation and an example? I would like to cut a release.

If not, we can wait until `v0.4.1`.

Reply:
I've updated docs, but I'm not sure what is a good example here, so I didn't add examples.

Reply:
That's fine, no need for example. I meant an example like "put barriers before and after the `nop` sled" if the exact number of `nop`s between I/O access is important".

In any case, merging this, thank you!